Mixin/Mixout Optimization for simple pipelines

ranj063 opened the pull request:
Includes the patch from #7792 

For a pipeline like: host-copier->gain->mixin->mixout->gain->dai-copier
The original MCPS (Measured on TGL chromebook) is:
```
            host-copier.2.playback,   comp:0 0x40000,     3.46,     6.89
                          gain.5.1,   comp:0 0x60000,     2.17,     9.22
                         mixin.5.1,   comp:0 0x20000,     2.62,     5.94
                        mixout.6.1,   comp:1 0x30000,     1.89,     2.42
                          gain.6.1,   comp:1 0x60001,     2.22,    16.57
 dai-copier.SSP.NoCodec-2.playback,   comp:1 0x40001,     6.05,     6.41
                                  ,                 ,    18.42,    47.44

```
And the optimized performance is:
```
                              Name,        Mtrace id, avg MCPS, max MCPS
            host-copier.2.playback,   comp:0 0x40000,     3.42,     6.86
                          gain.5.1,   comp:0 0x60000,     2.19,     9.24
                         mixin.5.1,   comp:0 0x20000,     0.80,     0.82
                        mixout.6.1,   comp:1 0x30000,     2.14,     2.44
                          gain.6.1,   comp:1 0x60001,     2.19,     9.34
 dai-copier.SSP.NoCodec-2.playback,   comp:1 0x40001,     5.90,    13.75
                                  ,                 ,    16.64,    42.46

```

andrula-song commented on src/audio/module_adapter/module_adapter.c:
```diff
@@ -825,6 +871,67 @@ static int module_adapter_simple_copy(struct comp_dev *dev)
+		struct comp_buffer *source;
+
+		source = list_first_item(&source_dev->bsource_list, struct comp_buffer,
+					 sink_list);
```
can we add a mod->source = source here. In case we may have pipeline like copier-> volume(bypass) -> mixin(bypass) ->mixout. If always take the last component's source as source may lead to unexpected error.

andrula-song replied:
actually, I think if there is only one mixin and one mixout, then the mixin and mixout both can be "bypass". and since we trigger processing from pipeline start to end, I don't think we need a loop here, if we always update the mod->source is its last component is bypass component.

andrula-song commented on src/audio/mixin_mixout/mixin_mixout.c:
```diff
@@ -482,14 +494,53 @@ static int mixout_process(struct processing_module *mod,
+
+			bytes = frames * audio_stream_frame_bytes(output_buffers[0].data);
+			input_buffers[0].consumed = bytes;
+			output_buffers[0].size = bytes;
```
can we set mod->dev->source_bypass_capable = true; and store mod->source & mod->sink  here? transfer pointer is faster than audio stream copy.

lyakh commented on src/audio/mixin_mixout/mixin_mixout.c:
```diff
@@ -482,14 +494,53 @@ static int mixout_process(struct processing_module *mod,
+		 * buffer into the sink buffer. Otherwise, continue on to the normal mixout
+		 * processing below.
+		 */
+		if (source_buffer->sink != mod->dev) {
```
this looks suspicious. If we add generic bypass mechanism, then individual modules shouldn't need to identify such cases, it should be transparent to them. Or maybe we should always call module's `.process()` method and those modules, that can and should be bypassed can indicate a short-circuit condition in which case the module adapter calls the next `.process()` with the old input?

ranj063 replied:
same problem with not all modules not using the module adapter.

ranj063 committed:
mixin_mixout: Optimize mixin_mixout

The current implementation of mixin/mixout with the mixin performing the
mixing and writing out to the mixout's sink buffer is not well suited to
be used with the module adapter.

So change the implementation to perform simple audio stream copy at the
mixin when it has multiple sinks and always perform the mixing operation
in the mixout process callback. When a mixin only has a single sink, its
processing can be bypassed altogether to optimize the performance.

This change removes support for channel remapping and can be added back
later if needed.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

andrula-song commented on src/audio/mixin_mixout/mixin_mixout.c:
```diff
@@ -482,14 +494,53 @@ static int mixout_process(struct processing_module *mod,
+		}
+
+		frames = audio_stream_avail_frames_aligned(input_buffers[0].data,
+							   output_buffers[0].data);
```
from your code line 814:
	frames = audio_stream_avail_frames_aligned(&source_c->stream, &sink_c->stream);
	mod->input_buffers[0].size = frames;
audio_stream_avail_frames_aligned return the MIN(avail, free).

ranj063 committed:
audio: component: Add new flags for module process bypass

In preparation for making the mixin module skip processingi when it has
only 1 sink, add a couple of new flags in struct comp_dev that will be
used to identify it as bypass capable and enable bypass.

When a module identifies as bypass_capable and bypass is enabled,
it's sink modules will skip their source buffers and directly use the
bypassed module's source buffers for processing.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
[TEST] check source dev state as well

just a test

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
[TEST] do not enable bypass in mixin

just a test

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

lyakh reviewed (commented):
not a real review (yet), just a couple of questions, that occurred to me
